### ISSUE TITLE
Incorrect encoding set for Norwegian Language

### ISSUE THREAD

**Issue by thohan88** [rstudio/rstudio#12668]
### System details

    RStudio Edition : Desktop
    RStudio Version : "Cherry Blossom" Daily (7f5869fc, 2023-02-01) for Windows
    OS Version      : Microsoft Windows 10 Enterprise (10.0.19044 Build 19044) 
    R Version       : R 4.2.2

### Steps to reproduce the problem
Starting up RStudio (either daily build or stable) with Norwegian locale on Windows will give a warning: 
```
Warning message:
In Sys.setlocale("LC_CTYPE", ctype) :
  OS reports request to set locale to "Norwegian Bokmål_Norway.utf8" cannot be honored
```
Attempting to write a character will halt the console:
![image](https://user-images.githubusercontent.com/16469189/216963481-6d49c402-3591-47ca-b98a-ee95b3f620db.png)

This is not an issue with RGui which appears to handle this just fine:
![image](https://user-images.githubusercontent.com/16469189/216963750-f6530cbf-61f0-4fb5-9852-057f9e7ead32.png)

It also causes problems for directory listings in RStudio, e.g.
![image](https://user-images.githubusercontent.com/16469189/216964826-4fa96442-f0fb-43c8-be06-63e054c94700.png)
![image](https://user-images.githubusercontent.com/16469189/216964903-ecb5a519-23a4-4590-8975-59b5e93d6250.png)

### Describe the problem in detail
The consequence is that scripts can't include Norwegian characters such as "æ", "ø" and "å".  Scripts including these characters can't be sourced even if the characters are only used in comments. This appears to be a widespread problem for Norwegian users as shown in the following recent issues: [1](https://community.rstudio.com/t/scandinavian-characters-in-rstudio/139796), [2](https://community.rstudio.com/t/rstudio-triggers-wrong-locale/137503) [3](https://stackoverflow.com/questions/72287390/sys-getlocale-fails-when-called-from-rstudio) and [4](https://stackoverflow.com/questions/75172826/rstudio-fails-to-find-right-locale-settings-at-startup)

### Describe the behavior you expected
Setting the system locale to `nb-NO.UTF-8` manually appears to fix the problem:
```
Sys.setlocale("LC_ALL", "nb-NO.UTF-8")
```
![image](https://user-images.githubusercontent.com/16469189/216965698-8543bc73-b3a5-4f75-8952-6218b5b068fd.png)

- [x] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [x] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [x] If I am reporting an RStudio crash, I have included a [diagnostics report](https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [x] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by kevinushey** [comment]
I don't see an issue when I try to manually set the locale on my Windows machine:

```
> Sys.setlocale("LC_ALL", "Norwegian Bokmål_Norway.utf8")
[1] "LC_COLLATE=Norwegian Bokmål_Norway.utf8;LC_CTYPE=Norwegian Bokmål_Norway.utf8;LC_MONETARY=Norwegian Bokmål_Norway.utf8;LC_NUMERIC=C;LC_TIME=Norwegian Bokmål_Norway.utf8"
```

Are you able to "manually" set the locale here?

It's not clear to me where that call to `Sys.setlocale()` is coming from, though.

**Comment by thohan88** [comment]
Thanks for the quick response, @kevinushey. 

Weird. Running that command crashes RStudio.

![example](https://user-images.githubusercontent.com/16469189/217061613-0e40aed7-5d94-420b-b48a-ba2f6723e260.gif)

However, if I run it after my previous command, it seems to work. 
```
Sys.setlocale("LC_ALL", "nb-NO.UTF-8")
Sys.setlocale("LC_ALL", "Norwegian Bokmål_Norway.utf8")
```
![image](https://user-images.githubusercontent.com/16469189/217061345-fb224af6-bf64-4297-a3e7-a29cde959a15.png)

I suspect it could be due to the `å` in `Sys.setlocale("LC_ALL", "Norwegian Bokmål_Norway.utf8")` since I'm unable to use it in a string in the console.

**Comment by EirikTengesdal** [comment]
I have the same issue, and have experienced it at least since I upgraded R to version 4.2.1. Downgrading to R version 4.1 has previously solved the problem for me.

I wonder whether the problem might be related to the major changes to R 4.2, presented here: "[R 4.2 for Windows will support UTF-8 as native encoding, which will be a major improvement in encoding support, allowing Windows R users to work with international text and data.](https://blog.r-project.org/2021/12/07/upcoming-changes-in-r-4.2-on-windows/)"?

## TL;DR:
My suggestion is to manually add to one's `.Rprofile.site` (apparently solving all related problems, also when some package command resets the locale to default/`C`):
```
.First <- function() {
   Sys.setlocale("LC_ALL", "nb-NO.UTF-8")
   }
```

### Problem description
I experience the same problems as @thohan88. I will therefore not repeat identical information below.

I have attempted two main workarounds, and they both seem to work as of now.

### System 
```
RStudio Edition : Desktop
RStudio Version : RStudio 2022.12.0+353 "Elsbeth Geranium" Release (7d165dcfc1b6d300eb247738db2c7076234f6ef0, 2022-12-03) for Windows
OS Version      : Microsoft Windows 10 Enterprise (10.0.19044 Build 19044) 
R Version       : R 4.2.2 (2022-10-31 ucrt)
```
### Workaround no. 1: Possible solution
In line with many suggestions, I ran the command `Sys.setlocale("LC_ALL")` (first without providing the `locale` specification (cf. the [default](https://stat.ethz.ch/R-manual/R-devel/library/base/html/locales.html)), which returns: `[1] "LC_COLLATE=Norwegian Bokmål_Norway.utf8;LC_CTYPE=Norwegian Bokmål_Norway.utf8;LC_MONETARY=Norwegian Bokmål_Norway.utf8;LC_NUMERIC=C;LC_TIME=Norwegian Bokmål_Norway.utf8"`.

All base commands I have tested works fine with this solution, irrespective of how many times I enter UTF-8 characters. However, when I run certain package commands, the locale somehow resets to `C`. This is reproducible from the second time it is run with, e.g., ggplot2::ggplot() iff new input contains UTF-8 characters:

```
ggplot2::ggplot()+
  ggplot2::labs(title = "Test with Å")
```
The first time this command is run, it correctly outputs a plot with the title _Test with Å_.

But rerunning the same command results in a failure (and the code must be run twice (or something else), as it expects more code after running the selected code containing a UTF-8 character). On the other hand, removing any UTF-8 character will result in a correct plot.

![Test with Å](https://user-images.githubusercontent.com/82897707/219757475-6ad34160-1b5a-42c0-bb44-5664fb26df77.png)

Rerunning the command `Sys.setlocale("LC_ALL")` solves the problem until next time the locale is reset.

This workaround is really not optimal as it requires the user to know when certain commands resets the locale, and many of my scripts that worked completely fine on earlier R versions would now need many `Sys.setlocale("LC_ALL")` insertions to prevent errors.

(HOWEVER: See workaround no. 2 below. Running the command `Sys.setlocale("LC_ALL", "nb-NO.UTF-8")` seems to enable also workaround no. 1 as a viable option.)

### Workaround no. 2: Possible solution
I first attempted to append the following code to my `.Rprofile.site`, based on a suggestion from [Stack Overflow](https://stackoverflow.com/questions/39340185/how-to-set-the-default-language-of-date-in-r):
```
.First <- function() {
   Sys.setlocale("LC_ALL")
   }
```
This produces the same warning at startup as without any such appended code to the profile:
```
Warning message:
In Sys.setlocale("LC_CTYPE", ctype) :
  OS reports request to set locale to "Norwegian Bokmål_Norway.utf8" cannot be honored
```
The same happens when I try `Sys.setlocale("LC_ALL", "Norwegian Bokmål_Norway.utf8")`.

But I just now tested @thohan88's hypothesis that the UTF-8 character `å` in the locale name string might be responsible for the error. By specifying `"nb-NO.UTF-8"` instead of `"Norwegian Bokmål_Norway.utf8"` in my `.Rprofile.site`, I seem to have solved the problem:
```
.First <- function() {
   Sys.setlocale("LC_ALL", "nb-NO.UTF-8")
   }
```
That is, at RStudio startup, I still get the warning message `In Sys.setlocale("LC_CTYPE", ctype) : OS reports request to set locale to "Norwegian Bokmål_Norway.utf8" cannot be honored`, but running the command `Sys.getlocale()` directly after startup confirms that the locale is correct after all:

![Locale](https://user-images.githubusercontent.com/82897707/219760591-41c46601-ce21-4fd3-b01c-3f8210e53db7.png)

Luckily, this also seems to fix the problem that occurred with the other manual fix option (i.e.,  `"Norwegian Bokmål_Norway.utf8"`), where certain ggplot2-commands that involve processing UTF-8 symbols result in resetting the locale to `C`.

## Conclusion
Thus I think @thohan88 is right in the suspicion that it is `å` that is the culprit in this case, and it seems most likely related to base R character string parsing (at least in conjunction with `Sys.setlocale()`) that, at startup, does not support UTF-8 encoding.

This entails a prediction that all users that default to a locale that contains a character string with at least one UTF-8 encoded character will face the same problem. However, it seems that several other people are affected by the same or some related problem than the ones with Scandinavian locales (see, e.g., [here](https://community.rstudio.com/t/what-is-the-state-of-utf-8-support-on-windows-10-with-r-4-0/76652/2) and [here](https://stackoverflow.com/questions/62726261/utf-8-support-in-r-on-windows)).

In any case, I really hope the issue will be addressed and fixed, since it disturbs the daily workflow if one does not put in place these measures that are not readily known nor available to users.

**Comment by szeestraten** [comment]
@kevinushey, @ronblum, apologies for nagging, but are there any other troubleshooting steps we can do or info we should collect to help?

In RStudio with R 4.2.3 and R 4.3.0, the locale defaults to `C` for an unknown reason.
Outside RStudio, the locale defaults to `Norwegian Bokmål_Norway.utf8` as expected.
R 4.1.3 works and has the same locale `Norwegian Bokmål_Norway.1252` for both.

```
Rstudio Version: 2023.03.0+386
```

#### Windows locale
```
PS C:\Users\sandor.zeestraten> Get-WinSystemLocale

LCID             Name             DisplayName
----             ----             -----------
1044             nb-NO            Norwegian, Bokmål (Norway)
```

#### sessionInfo from RStudio
```
> sessionInfo()
R version 4.3.0 (2023-04-21 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19045)

Matrix products: default

locale:
[1] C

time zone: Europe/Oslo
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_4.3.0 tools_4.3.0   
```

```
> sessionInfo()
R version 4.2.3 (2023-03-15 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19045)

Matrix products: default

locale:
[1] C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_4.2.3 tools_4.2.3
```

```
> sessionInfo()
R version 4.1.3 (2022-03-10)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19045)

Matrix products: default

locale:
[1] LC_COLLATE=Norwegian Bokmål_Norway.1252  LC_CTYPE=Norwegian Bokmål_Norway.1252    LC_MONETARY=Norwegian Bokmål_Norway.1252
[4] LC_NUMERIC=C                             LC_TIME=Norwegian Bokmål_Norway.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_4.1.3 tools_4.1.3   
```

**Comment by kevinushey** [comment]
I'm trying to figure out where R / RStudio is making this call to `Sys.setlocale()`, but am having no luck. To confirm, are you calling `Sys.setlocale()` yourself somewhere, or is that happening automatically for you behind the scenes?

**Comment by kevinushey** [comment]
Do you get any useful information if you include the following in your `~/.Rprofile`?

```
options(warn = 2)
options(error = rlang::trace_back)
```

Or is the warning being produced before the `.Rprofile` is run?

**Comment by EirikTengesdal** [comment]
Thank you, @kevinushey. This is happening automatically for me behind the scenes. Lately, even the workaround I suggested in my post from 17 February 2023 (i.e., adding `Sys.setlocale("LC_ALL", "nb-NO.UTF-8")` to my `.Rprofile`) occasionally fails.

It might be coincidental, but this has happened when executing several R code chunks in R Markdown. I think running the code and knitting both fails, so I had to repeat the line `Sys.setlocale("LC_ALL")` before all problematic lines. For instance, this happened when I used pipes and dplyr, e.g., `df %>% filter(County=="Trøndelag")` (`ø` is UTF-8).

If I remember correctly, saving a dataframe following a `dplyr::mutate()` operation with strings containing `æøå` sometimes also causes undesirable effects (viewing the string in the resulting dataframe in RStudio showed, e.g., `\u00F8` instead of `ø`).

I wonder whether the problem could be caused by a package or some dependencies. I have forgotten the source now, but vaguely remember reading somewhere that a specific package conditionally runs `Sys.setlocale()` or something like that, yielding a reset to the default locale `C`. I can try to find it if you think it will be helpful.

In any case, I will include your suggestion in my `.Rprofile`, test it and come back to you ASAP. I do not think the warning is produced before the `.Rprofile` is run.

**Comment by kevinushey** [comment]
Thanks -- it would be helpful to know when exactly the warning is produced; if I can reproduce the issue locally I'll have a better chance of resolving the issue.

**Comment by szeestraten** [comment]
@kevinushey, thanks for the update.

Just to clarify, in my setup the locale defaults to `C`. It is a plain install of R and RStudio and I am not normally calling `Sys.setlocale()` besides for the troubleshooting here.
Note that I do **not** get the warning message (`OS reports request to set locale to "Norwegian Bokmål_Norway.utf8" cannot be honored`) as experienced by the @thohan88 and @EirikTengesdal.

Unfortunately, the `options` do not produce any more info in my case.
I wonder if you'd be able to reproduce it if with setting Norwegian as the locale in Windows.

I also tried a few different combinations of locales in `.Rprofile` if that helps (first part is `.Rprofile` snippet, second part is console output):

```
Sys.getlocale()
Sys.setlocale()
Sys.getlocale()

[1] "LC_COLLATE=Norwegian Bokmål_Norway.utf8;LC_CTYPE=Norwegian Bokmål_Norway.utf8;LC_MONETARY=Norwegian Bokmål_Norway.utf8;LC_NUMERIC=C;LC_TIME=Norwegian Bokmål_Norway.utf8"
[1] "LC_COLLATE=Norwegian Bokmål_Norway.utf8;LC_CTYPE=Norwegian Bokmål_Norway.utf8;LC_MONETARY=Norwegian Bokmål_Norway.utf8;LC_NUMERIC=C;LC_TIME=Norwegian Bokmål_Norway.utf8"
[1] "LC_COLLATE=Norwegian Bokmål_Norway.utf8;LC_CTYPE=Norwegian Bokmål_Norway.utf8;LC_MONETARY=Norwegian Bokmål_Norway.utf8;LC_NUMERIC=C;LC_TIME=Norwegian Bokmål_Norway.utf8"
> Sys.getlocale()
[1] "C"
```

```
Sys.getlocale()
Sys.setlocale("LC_ALL", "Norwegian Bokmål_Norway.utf8")
Sys.getlocale()

[1] "LC_COLLATE=Norwegian Bokmål_Norway.utf8;LC_CTYPE=Norwegian Bokmål_Norway.utf8;LC_MONETARY=Norwegian Bokmål_Norway.utf8;LC_NUMERIC=C;LC_TIME=Norwegian Bokmål_Norway.utf8"
[1] "LC_COLLATE=Norwegian Bokmål_Norway.utf8;LC_CTYPE=Norwegian Bokmål_Norway.utf8;LC_MONETARY=Norwegian Bokmål_Norway.utf8;LC_NUMERIC=C;LC_TIME=Norwegian Bokmål_Norway.utf8"
[1] "LC_COLLATE=Norwegian Bokmål_Norway.utf8;LC_CTYPE=Norwegian Bokmål_Norway.utf8;LC_MONETARY=Norwegian Bokmål_Norway.utf8;LC_NUMERIC=C;LC_TIME=Norwegian Bokmål_Norway.utf8"
> Sys.getlocale()
[1] "C"
```

```
Sys.getlocale()
Sys.setlocale("LC_ALL", "nb-NO.UTF-8")
Sys.getlocale()

[1] "LC_COLLATE=Norwegian Bokmål_Norway.utf8;LC_CTYPE=Norwegian Bokmål_Norway.utf8;LC_MONETARY=Norwegian Bokmål_Norway.utf8;LC_NUMERIC=C;LC_TIME=Norwegian Bokmål_Norway.utf8"
[1] "LC_COLLATE=nb-NO.UTF-8;LC_CTYPE=nb-NO.UTF-8;LC_MONETARY=nb-NO.UTF-8;LC_NUMERIC=C;LC_TIME=nb-NO.UTF-8"
[1] "LC_COLLATE=nb-NO.UTF-8;LC_CTYPE=nb-NO.UTF-8;LC_MONETARY=nb-NO.UTF-8;LC_NUMERIC=C;LC_TIME=nb-NO.UTF-8"
> Sys.getlocale()
[1] "LC_COLLATE=nb-NO.UTF-8;LC_CTYPE=nb-NO.UTF-8;LC_MONETARY=nb-NO.UTF-8;LC_NUMERIC=C;LC_TIME=nb-NO.UTF-8"
```

```
.First <- function() {
  Sys.setlocale()
}
Sys.getlocale()

[1] "LC_COLLATE=Norwegian Bokmål_Norway.utf8;LC_CTYPE=Norwegian Bokmål_Norway.utf8;LC_MONETARY=Norwegian Bokmål_Norway.utf8;LC_NUMERIC=C;LC_TIME=Norwegian Bokmål_Norway.utf8"
> Sys.getlocale()
[1] "C"
```

```
.First <- function() {
  Sys.setlocale("LC_ALL", "Norwegian Bokmål_Norway.utf8")
}
Sys.getlocale()

[1] "LC_COLLATE=Norwegian Bokmål_Norway.utf8;LC_CTYPE=Norwegian Bokmål_Norway.utf8;LC_MONETARY=Norwegian Bokmål_Norway.utf8;LC_NUMERIC=C;LC_TIME=Norwegian Bokmål_Norway.utf8"
> Sys.getlocale()
[1] "C"
```

```
.First <- function() {
  Sys.setlocale("LC_ALL", "nb-NO.UTF-8")
}
Sys.getlocale()

[1] "LC_COLLATE=Norwegian Bokmål_Norway.utf8;LC_CTYPE=Norwegian Bokmål_Norway.utf8;LC_MONETARY=Norwegian Bokmål_Norway.utf8;LC_NUMERIC=C;LC_TIME=Norwegian Bokmål_Norway.utf8"
> Sys.getlocale()
[1] "LC_COLLATE=nb-NO.UTF-8;LC_CTYPE=nb-NO.UTF-8;LC_MONETARY=nb-NO.UTF-8;LC_NUMERIC=C;LC_TIME=nb-NO.UTF-8"
```

**Comment by huftis** [comment]
Just two quick comments:

This bug only occurs when the display language in Windows is set to Norwegian Bokmål. Specifically, it does *not* happen when the display language is set to Norwegian Nynorsk (which have identical regional settings as Norwegian Bokmål). So it’s probably the letter `å` that’s at the root cause of the problem (wrong encoding or double-encoding when reading/storing the name of the locale?).

You don’t need an R *command* for a workaround. You only have to set

    LC_ALL="nb_NO.utf8"

in your .Renviron file.

**Comment by ohjakobsen** [comment]
We are experiencing the same issue on Windows 10 (build 19044) with both R 4.2 and 4.3 and the system language set to Norwegian Bokmål. RGui and R.exe will report correct locale, but RStudio resets locale to C. Changing the system language to Norwegian Nynorsk solves the issue with RStudio. To fix the problem with Norwegian Bokmål, we have added the following to `.Rprofile` which seems to work.
```r
.First <- function() {
  Sys.setlocale('LC_ALL', 'nb-NO.utf-8')
}
```
